### PR TITLE
adds PowerShell argument tab completion for Paket-Add

### LINKED ISSUE #873: Paket.PowerShell Argument Tab Completion
An obvious enhancement for the new PowerShell.Paket is tab completion for package names and then versions across the commands. I have to google some more to figure out how to do it. It is next on my radar.

seemed helpful:
image

### PR DESCRIPTION
**Pull request by ctaggart**
fix #873
fix exception handling
add some positional params

image

image

image
The list of versions should be sorted the other way. See #886.

### CODE REVIEW

**Comment by ctaggart** [comment]
Last issue before this is mergable is this is causing things to lock up.
``` powershell
Pkaet-Add -NuGet
```

**Comment by ctaggart** [comment]
This is good to go.

**Comment by ctaggart** [comment]
I'm going to post more screenshots. I'm pretty happy with this. I'm looking forward to announcing `Paket.PowerShell` soon.

image

image

As soon as get this flipped: :)
image

**Comment by forki** [comment]
Will fix later today, but needs time until the chocolatey people release it

**Comment by ctaggart** [comment]
In my blog, I'll talk about adding other chocolatey sources like the NuGet Gallery to stay up-to-date.

``` powershell
choco source list
choco source add -n=nuget -s=https://www.nuget.org/api/v2
```

**Comment by forki** [comment]
Nice. And I read they can also white-list tools. Maybe they will white-list
paket one day.

File: src/Paket.PowerShell/PowerShell.fs [hunk 1]
```diff
@@ -25,7 +25,10 @@ module PaketPs =
         async {
             try
                 do! Async.SwitchToNewThread()
-                do! computation
+                try
```
Comment (ctaggart) [inline src/Paket.PowerShell/PowerShell.fs]: Without the `try with` here, an exception kills the started thread, which kills the powershell thread which kills powershell. It is important.

Comment (forki) [inline src/Paket.PowerShell/PowerShell.fs]: Could you please add this as a comment in the code there.

File: src/Paket.PowerShell/ArgumentTabCompletion.ps1 [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        createCompletionResult $_ $_ $_ | write
+    }
+}
+
+# create and add $global:options to the list of completers
+# http://www.powertheshell.com/dynamicargumentcompletion/
+if (-not $global:options) { $global:options = @{CustomArgumentCompleters = @{};NativeArgumentCompleters = @{}}}
+
+$global:options['CustomArgumentCompleters']['Paket-Add:NuGet'] = $findPackages
+$global:options['CustomArgumentCompleters']['Paket-Add:Version'] = $findPackageVersions
```
Comment (ctaggart) [inline src/Paket.PowerShell/ArgumentTabCompletion.ps1]: More argument tab completers can be added following this pattern. It must be done in PowerShell. The output of another command is used to do the tab completion.

**Pull request merged**